ipld/retriever.go

```diff
@@ -0,0 +1,178 @@
+package ipld
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipfs/go-merkledag"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var log = logging.Logger("ipld")
+
+// RetrieveQuadrantTimeout limits the time for retrieval of a quadrant
+// so that Retriever can retry another quadrant.
+var RetrieveQuadrantTimeout = time.Minute * 5
```

Comment:
What is this timeout based on?

Reply:
I didn't put too much thought into this but rn my thoughts are:
* This was just the first number that came into my mind after Mustafa calculated the possible amount of proofs that need to be downloaded with 4mb blocks, thus roundtrips. 
* This number should cover some discovery waiting time. 
* Setting this number with some leg room ensures we dont cancel the operation too early if the operation tends to be long
* To have a better timeout we should make some real-world measurements.

Reply:
An issue: https://github.com/celestiaorg/celestia-node/issues/639